pkg/component/controller/cplb_reconciler.go

```diff
@@ -0,0 +1,175 @@
+		return fmt.Errorf("Failed to get clientset. Unable to : %w", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	r.watchCancelFunc = cancel
+	go r.watchAPIServers(ctx)
+
+	return nil
+}
+
+func (r *CPLBReconciler) Stop() {
+	r.log.Info("Stopping CPLB reconciler")
+	r.watchCancelFunc()
+}
+
+func (r *CPLBReconciler) watchAPIServers(ctx context.Context) {
+	// Before starting check that the API is actually responding
+	for {
+		// watch the "kubernetes" endpoints in the "default" namespace
+		watcher, err := r.clientset.CoreV1().Endpoints("default").Watch(ctx, metav1.ListOptions{
+			FieldSelector: "metadata.name=kubernetes",
+		})
+		if err != nil {
+			r.log.Errorf("Failed to watch kubernetes endpoints. Retrying in 5 seconds : %v", err)
+			time.Sleep(5 * time.Second)
+			continue
+		}
+
+		r.log.Info("Kubernetes endpoint watcher started")
+	loop:
+		for {
+			select {
+			case event, ok := <-watcher.ResultChan():
+				if !ok {
+					r.log.Warn("Kubernetes endpoint watcher closed, retrying to open it")
+					time.Sleep(5 * time.Second)
+					break loop
+				}
+				switch event.Type {
+				case watch.Added, watch.Modified:
+					endpoint, ok := event.Object.(*corev1.Endpoints)
+					if ok {
+						r.log.Info("Kubernetes endpoint was updated")
+						r.maybeUpdateIPs(endpoint)
+					}
+				case watch.Deleted:
+					r.log.Error("Kubernetes endpoint was deleted, this is not supposed to happen. Ignoring the event")
+				case watch.Error:
+					r.log.Errorf("Kubernetes endpoint watch error: %v", event.Object)
+					time.Sleep(5 * time.Second)
+					break loop
+				}
+			case <-ctx.Done():
+				watcher.Stop()
+				r.log.Info("Stopped watching kubernetes endpoints")
+				return
+			}
+		}
+	}
+}
+
+// maybeUpdateIPs updates the list of IP addresses if the new list has
+// different addresses
+func (r *CPLBReconciler) maybeUpdateIPs(endpoint *corev1.Endpoints) {
+	newAddresses := []string{}
+	for _, subset := range endpoint.Subsets {
+		for _, addr := range subset.Addresses {
+			newAddresses = append(newAddresses, addr.IP)
+		}
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if len(newAddresses) != len(r.addresses) {
+		r.updateAddresses(newAddresses)
+		return
+	}
+
+	// Subsets are not guaranteed to be ordered, so we need to compare
+	// the lists
+	for _, addr := range newAddresses {
+		if !slices.Contains(r.addresses, addr) {
+			r.updateAddresses(newAddresses)
+			return
+		}
+	}
+	for _, addr := range r.addresses {
+		if !slices.Contains(newAddresses, addr) {
+			r.updateAddresses(newAddresses)
+			return
+		}
+	}
+}
+
+func (r *CPLBReconciler) updateAddresses(newAddresses []string) {
+	r.addresses = newAddresses
+	r.log.Infof("Updated the list of IPs: %v", r.addresses)
+	select {
+	case r.updateCh <- true:
+	default:
+	}
+}
+
+// GetIPs gets a thread-safe copy of the current list of IP addresses
+func (r *CPLBReconciler) GetIPs() []string {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	return r.addresses
```

Comment:
Might be a good idea to clone the slice here, so that there's no surprises when callers start to modify it.

```suggestion
	return slices.Clone(r.addresses)
```